File: USACO/Cow Tours/cowtour.py
"""
ID: mjmande1
LANG: PYTHON3
TASK: cowtour
"""

# def distnace(x1, y1, x2, y2):
#     return ((x1 - x2) ** 2 + (y1 - y2) ** 2) ** 0.5



# with open("cowtour.in", "r") as fin, open("cowtour.out", "w") as fout:

#     locations = []
#     connections = []

#     N = int(fin.readline().strip())
#     for _ in range(N):
#         locations.append(tuple(map(int, fin.readline().strip().split())))
    
#     for y in range(N):
#         line = fin.readline().strip()
#         for x, c in enumerate(line):
#             if c == "1":
#                 connections.append((y, x))
    
#     print(locations)
#     print(connections)

#     graph = {}
#     for p1, p2 in connections:
#         if p1 not in graph:
#             graph[p1] = {}
#         if p2 not in graph:
#             graph[p2] = {}

#         graph[p1][p2] = distnace(locations[p1][0], locations[p1][1], locations[p2][0], locations[p2][1])
#         graph[p2][p1] = distnace(locations[p1][0], locations[p1][1], locations[p2][0], locations[p2][1])
        
    
    

#     fout.close()
#     fin.close()

import math

def solve(current, number):
    if visited[current] != -1:
        return
    visited[current] = number
    for i in range(N):
        if adjacent[current][i] == 1:
            solve(i, number)

def floyd(matrix):
    for i in range(N):
        for j in range(N):
            for k in range(N):
                matrix[j][k] = min(matrix[j][k], matrix[j][i]+matrix[i][k])
    return matrix

# set startTime to measure how long the program takes

# read input from file
with open('cowtour.in', 'r') as file:
    N = int(file.readline().strip())
    location = [[None]*2 for i in range(N)]
    adjacent = [[0]*N for i in range(N)]
    visited = [-1] * N
    for i in range(N):
        line = file.readline().split()
        location[i][0] = int(line[0])
        location[i][1] = int(line[1])
        
    for i in range(N):
        lineArray = file.readline().strip()
        for j in range(N):
            adjacent[i][j] = int(lineArray[j])

visited = [-1] * N
for i in range(N):
    solve(i, i)

adjacencyMatrix = [[float("inf")]*N for i in range(N)]
for i in range(N):
    for j in range(N):
        if i == j:
            adjacencyMatrix[i][j] = 0
        elif adjacent[i][j]==1:
            distance = math.hypot(location[j][0]-location[i][0], location[j][1]-location[i][1])
            adjacencyMatrix[i][j] = distance
        else:
            adjacencyMatrix[i][j] = float("inf")
            
# run Floyd-Warshall Algorithm
adjacencyMatrix = floyd(adjacencyMatrix)

minDistance = float("inf")
for i in range(N):
    for j in range(i+1, N):
        if visited[i] == visited[j]:
            continue
        distance = math.hypot(location[j][0]-location[i][0], location[j][1]-location[i][1])
        maximum = 0.0
        for k in range(N):
            if visited[k] != visited[i] and visited[k] != visited[j]:
                continue
            for m in range(k+1, N):
                if (visited[m] != visited[i] and visited[m] != visited[j]) or adjacencyMatrix[k][m] < maximum:
                    continue
                if visited[i] == visited[k]:
                    maximum = max(maximum, min(adjacencyMatrix[k][m], distance+adjacencyMatrix[k][i]+adjacencyMatrix[j][m]))
                else:
                    maximum = max(maximum, min(adjacencyMatrix[k][m], distance+adjacencyMatrix[k][j]+adjacencyMatrix[i][m]))
                    
                if maximum > minDistance:
                    break
        minDistance = min(minDistance, maximum)
           
with open('cowtour.out', 'w') as file:
    result = '{:.6f}'.format(minDistance).rstrip('0')
    if result[-1] == '.':
        result += '0'

    # add trailing 0s until length of decimal is 6
    while len(result.split('.')[1]) < 6:
        result += '0'

    file.write(result + '\n')

# print final time taken

File: USACO/Cow Tours/cowtour.cpp
/*
ID: mjmande1
LANG: C++
TASK: cowtour
*/

#include <bits/stdc++.h>
using namespace std;

int N;
vector<pair<int, int>> location;
vector<vector<int>> adjacent;
vector<int> visited;

double distance(int x1, int y1, int x2, int y2) {
    return sqrt(pow((x1 - x2), 2) + pow((y1 - y2), 2));
}

void solve(int current, int number) {
    if (visited[current] != -1){
        return;
    }
    visited[current] = number;
    for (int i = 0; i < N; i++) {
        if (adjacent[current][i] == 1){
            solve(i, number);
        }
    }
}

vector<vector<double>> floyd(vector<vector<double>>& matrix) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            for (int k = 0; k < N; k++) {
                matrix[j][k] = min(matrix[j][k], matrix[j][i] + matrix[i][k]);
            }
        }
    }
    return matrix;
}

int main() {
    freopen("cowtour.in", "r", stdin);
    freopen("cowtour.out", "w", stdout);

    cin >> N;
    location.resize(N);
    adjacent.assign(N, vector<int>(N));
    visited.assign(N, -1);

    for (int i = 0; i < N; i++){
        cin >> location[i].first >> location[i].second;
    }
    
    for (int i = 0; i < N; i++){
        string lineArray;
        cin >> lineArray;
        for (int j = 0; j < N; j++){
            adjacent[i][j] = lineArray[j] - '0';
        }
    }

    visited.assign(N, -1);
    for (int i = 0; i < N; i++){
        solve(i, i);
    }

    vector<vector<double>> adjacencyMatrix(N, vector<double>(N, INT_MAX));
    for (int i = 0; i < N; i++){
        for (int j = 0; j < N; j++) {
            if (i == j){
                adjacencyMatrix[i][j] = 0;
            } else if (adjacent[i][j] == 1){
                double d = distance(location[j].first, location[j].second, location[i].first, location[i].second);
                adjacencyMatrix[i][j] = d;
            } else {
                adjacencyMatrix[i][j] = INT_MAX;
            }
        }
    }
            
    // run Floyd-Warshall Algorithm
    adjacencyMatrix = floyd(adjacencyMatrix);

    double minDistance = INT_MAX;
    for (int i = 0; i < N; i++){
        for (int j = i+1; j < N; j++){
            if (visited[i] == visited[j]){
                continue;
            }
            double dist = distance(location[j].first, location[j].second, location[i].first, location[i].second);
            double maximum = 0.0;
            for (int k = 0; k < N; k++){
                if (visited[k] != visited[i] && visited[k] != visited[j]){
                    continue;
                }
                for (int m = k+1; m < N; m++){
                    if ((visited[m] != visited[i] && visited[m] != visited[j]) || adjacencyMatrix[k][m] < maximum){
                        continue;
                    }
                    if (visited[i] == visited[k]){
                        maximum = max(maximum, min(adjacencyMatrix[k][m], dist + adjacencyMatrix[k][i] + adjacencyMatrix[j][m]));
                    } else {
                        maximum = max(maximum, min(adjacencyMatrix[k][m], dist + adjacencyMatrix[k][j] + adjacencyMatrix[i][m]));
                    }
                    
                    if (maximum > minDistance){
                        break;
                    }
                }
                if (maximum > minDistance){
                    break;
                }
            }
            minDistance = min(minDistance, maximum);
        }
    }

    printf("%.6f\n", minDistance);
    return 0;
}
